pgjdbc/src/main/java/org/postgresql/util/PasswordUtil.java

```diff
@@ -0,0 +1,218 @@
+    if (password == null) {
+      throw new NullPointerException("password is null");
+    }
+    if (iterations <= 0) {
+      throw new IllegalArgumentException("iterations must be greater than zero");
+    }
+    if (salt.length == 0) {
+      throw new IllegalArgumentException("salt length must be greater than zero");
+    }
+    try {
+      String passwordText = String.valueOf(password);
+      byte[] saltedPassword = ScramFunctions.saltedPassword(ScramMechanisms.SCRAM_SHA_256,
+          StringPreparations.SASL_PREPARATION, passwordText, salt, iterations);
+      byte[] clientKey = ScramFunctions.clientKey(ScramMechanisms.SCRAM_SHA_256, saltedPassword);
+      byte[] storedKey = ScramFunctions.storedKey(ScramMechanisms.SCRAM_SHA_256, clientKey);
+      byte[] serverKey = ScramFunctions.serverKey(ScramMechanisms.SCRAM_SHA_256, saltedPassword);
+
+      return "SCRAM-SHA-256" //
+        + "$" + iterations //
+        + ":" + Base64.toBase64String(salt) //
+        + "$" + Base64.toBase64String(storedKey) //
+        + ":" + Base64.toBase64String(serverKey);
+    } finally {
+      Arrays.fill(password, ' ');
+    }
+  }
+
+  /**
+   * Encode the given password for SCRAM-SHA-256 authentication using the default
+   * iteration count and a random salt.
+   *
+   * @param password The plain text of the user's password
+   * @return The text representation of the password encrypted for SCRAM-SHA-256
+   *         authentication
+   */
+  public static String encodeScramSha256(char[] password) {
+    if (password == null) {
+      throw new NullPointerException("password is null");
+    }
+    try {
+      SecureRandom rng = getSecureRandom();
+      byte[] salt = rng.generateSeed(DEFAULT_SALT_LENGTH);
+      return encodeScramSha256(password, DEFAULT_ITERATIONS, salt);
+    } finally {
+      Arrays.fill(password, ' ');
+    }
+  }
+
+  /**
+   * Encode the the given password for use with md5 authentication. The PostgreSQL
+   * server uses the username as the per-user salt so that must also be provided.
+   * The return value of this method is the literal text that may be used when
+   * creating or modifying a user with the given password without the surrounding
+   * single quotes.
+   *
+   * @param user     The username of the database user
+   * @param password The plain text of the user's password
+   * @return The text representation of the password encrypted for md5
+   *         authentication.
+   */
+  public static String encodeMd5(String user, char[] password) {
+    if (password == null) {
+      throw new NullPointerException("password is null");
+    }
+    try {
+      byte[] userBytes = user.getBytes(StandardCharsets.UTF_8);
+      String passwordText = String.valueOf(password);
+      byte[] passwordBytes = passwordText.getBytes(StandardCharsets.UTF_8);
+      final MessageDigest md = MessageDigest.getInstance("MD5");
+
+      md.update(passwordBytes);
```

Comment:
```suggestion
      byte[] passwordBytes = StandardCharsets.UTF_8.encode(CharBuffer.wrap(password)).array();
      try {
        MessageDigest md = MessageDigest.getInstance("MD5");

        md.update(passwordBytes);
      } finally {
        Arrays.fill(passwordBytes, (byte) 0);
      }
```

Reply:
Nope that doesn't work. Was going nuts trying to figure out why as it broke a bunch of the md5 tests. Turns out the `.array()` call returns the backing array of the ByteBuffer which is not necessarily the same size as the encoded password. In the tests it ends up being 38 bytes (extra zeros at the end) vs. 35 bytes and the extra zeros get included in the digest.

Going to try something similar to see if we can avoid the extra String in between. I think we can write the ByteBuffer directly to the MessageDigest then clear it.

Reply:
Yes the ByteBuffer worked when directly writing it to the MessageDigest. That way it's aware of the contents and not the entire backing array.